Userland/Libraries/LibWeb/Painting/StackingContext.cpp

```diff
@@ -117,6 +117,8 @@ void StackingContext::paint_descendants(PaintContext& context, Layout::Node cons
             if (child_is_inline_or_replaced) {
                 paint_node(child, context, PaintPhase::Background);
                 paint_node(child, context, PaintPhase::Border);
+                if (child.display().is_table_inside() && child.computed_values().border_collapse() == CSS::BorderCollapse::Separate)
+                    paint_table_borders(context, child);
```

Comment:
hm, it looks like `paint_table_borders` should happen somewhere inside `paint_node` and only for `PaintPhase::Border`.

Reply:
The phase is called `BackgroundAndBordersForInlineLevelAndReplaced` and this is an inline table, which is explicitly skipped during the `BackgroundAndBorders`. Why would we move it elsewhere?

Reply:
why can't border be painted during `PaintPhase::Border` for inline table because it seems to exist https://github.com/SerenityOS/serenity/blob/master/Userland/Libraries/LibWeb/Painting/StackingContext.cpp#L138. that wouldn't be correct?

Reply:
Where would we insert the call, though? `paint_node` is merely a wrapper for calling `paint` on a `Paintable`.

Reply:
into `PaintableBox::paint()`?

Reply:
The problem with doing that (or related approaches) is double-painting in collapsed borders mode. Since we create a bordered box for every cell, we would get some borders painted from that in addition to the borders we paint with the dedicated `paint_table_borders`. In `Paintable` it'd be a bit difficult to coordinate that, since the `border-collapse` property isn't accessible from there (nor shouldn't be).

Your larger point that painting table borders should funnel through the regular `paint_node` and `Border` phase is well taken, but in order to achieve that we might need to tweak how we create the paintable boxes for a table. This is beyond the current PR, but addressing it (and other rendering order issues, including backgrounds) is on my radar.

Reply:
hmm, I don't really understand what do you mean by double painting. You can access `border-collapse` property in `Paintable` or `PaintableBox` using `layout_node()->computed_values().border_collapse()` and I don't understand what is wrong with that.

Reply:
oh, I see what you mean. thank you for clarifying. it seems that in the future we would have to get rid of `paint_table_borders` and move painting into paintables that should bother about borders.

Reply:
Yes, `paint_table_borders` should probably be part of a `paint` for the entire table box. The current situation is a kludge which gives us respectably-looking borders, but it's far from ideal.